### PR TITLE
perf(client): stop exporting dmmf.schema

### PR DESCRIPTION
**Pull request by aqrln**
Closes https://github.com/prisma/client-planning/issues/17

### CODE REVIEW

File: packages/client/src/generation/utils/buildDMMF.ts [hunk 1]
```diff
@@ -1,18 +1,24 @@
+import { DMMF } from '@prisma/generator-helper'
 import lzString from 'lz-string'
 
+import { escapeJson } from '../TSClient/helpers'
+
 /**
  * Creates the necessary declarations to embed the generated DMMF into the
  * generated client. It compresses the DMMF for the data proxy engine.
```
Comment (millsp) [inline packages/client/src/generation/utils/buildDMMF.ts:8]:
```suggestion
 * generated client. It compresses the DMMF for the data proxy engine. For
 * data proxy, the full DMMF is exported, otherwise `schema` is removed.
```

File: packages/client/src/runtime/getPrismaClient.ts [hunk 2]
```diff
@@ -38,6 +40,7 @@ import { getRejectOnNotFound } from './utils/rejectOnNotFound'
 import { serializeRawParameters } from './utils/serializeRawParameters'
 import { validatePrismaClientOptions } from './utils/validatePrismaClientOptions'
 
+const P = Promise.resolve()
```
Comment (millsp) [inline packages/client/src/runtime/getPrismaClient.ts:43]: I think is not needed anymore.

File: packages/engine-core/src/binary/BinaryEngine.ts [hunk 3]
```diff
@@ -844,6 +845,26 @@ You very likely have the wrong "binaryTarget" defined in the schema.prisma file.
...
+    }
+    return this.getDmmfPromise
+  }
+
+  private async _getDmmf(): Promise<DMMF.Document> {
+    const prismaPath = await this.getPrismaPath()
+
+    const env = await this.getEngineEnvVars()
+
+    const result = await execa(prismaPath, ['cli', 'dmmf'], {
```
Comment (millsp) [inline packages/engine-core/src/binary/BinaryEngine.ts:860]:
```suggestion
    const result = await execa(prismaPath, ['--enable-raw-queries', 'cli', 'dmmf'], {
```

This is needed for it to work.

File: packages/engine-core/src/data-proxy/DataProxyEngine.ts [hunk 4]
```diff
@@ -88,6 +89,12 @@ export class DataProxyEngine extends Engine {
     } as GetConfigResult)
   }
 
+  getDmmf(): Promise<DMMF.Document> {
```
Comment (millsp) [inline packages/engine-core/src/data-proxy/DataProxyEngine.ts:92]: Maybe add a comment that this path should never be hit as it is handled upstream in `getPrismaClient`.

File: packages/engine-core/src/library/DefaultLibraryLoader.ts [hunk 5]
```diff
@@ -13,6 +13,7 @@ import { Library, LibraryLoader } from './types/Library'
 
 const debug = Debug('prisma:client:libraryEngine:loader')
 
+let library: Library | undefined // does not load twice in the same process
```
Comment (millsp) [inline packages/engine-core/src/library/DefaultLibraryLoader.ts:16]:
```suggestion
```

Let's remove this as it's unrelated.

File: packages/engine-core/src/library/LibraryEngine.ts [hunk 6]
```diff
@@ -91,10 +92,11 @@ export class LibraryEngine extends Engine {
     }
     this.libraryInstantiationPromise = this.instantiateLibrary()
 
-    initHooks()
+    initHooks(this)
```
Comment (millsp) [inline packages/engine-core/src/library/LibraryEngine.ts:95]: Lat's revert this as well.

File: packages/engine-core/src/library/LibraryEngine.ts [hunk 7]
```diff
@@ -489,28 +500,26 @@ You may have to run ${chalk.greenBright('prisma generate')} for your changes to
   }
 }
 
-function hookProcess(handler: string, exit = false) {
+function hookProcess(engine: LibraryEngine, handler: string, exit = false) {
```
Comment (millsp) [inline packages/engine-core/src/library/LibraryEngine.ts:503]: We can remove these unrelated changes in this function too.

File: packages/engine-core/src/library/DefaultLibraryLoader.ts [hunk 8]
```diff
@@ -30,7 +31,7 @@ export class DefaultLibraryLoader implements LibraryLoader {
     debug(`loadEngine using ${this.libQueryEnginePath}`)
     try {
       // this require needs to be resolved at runtime, tell webpack to ignore it
-      return eval('require')(this.libQueryEnginePath) as Library
+      return (library = library ?? (eval('require')(this.libQueryEnginePath) as Library))
```
[the remark this answers is not in the document]
Comment (aqrln) [reply packages/engine-core/src/library/DefaultLibraryLoader.ts:34]: Yes, absolutely. Unrelated changes not being cleaned up is the only reason the PR wasn't marked ready yet, I already made all the changes I wanted in the relevant code.

File: packages/engine-core/src/binary/BinaryEngine.ts (hunk 3 above)
Comment (aqrln) [reply packages/engine-core/src/binary/BinaryEngine.ts:860]: Thank you!

**Comment by aqrln** [comment]
TODO: move `Pick<DMMF.Document, 'datamodel' | 'mappings'>` type definition to runtime bundle and reference by name, or use raw `{ [K in 'datamodel' | 'mappings']: DMMF.Document[K] }`.

`Pick<T, K>` conflicts with potential model name `Pick`.

**Comment by aqrln** [comment]
TODO: investigate `missing-engine` tests failing due to ``Invalid `prisma.user.findMany()` invocation in ...`` not being present in the expected error message:

```
    - Snapshot  - 9
    + Received  + 1

    @@ -1,14 +1,6 @@
    -
    - Invalid `prisma.user.findMany()` invocation in
    - /client/src/__tests__/integration/errors/missing-engine-native-binaryTarget/library.test.ts:0:0
    -
    -   31 })
    -   32 
    -   33 await expect(async () => {
    - → 34   await prisma.user.findMany(
    -   Query engine library for current platform "TEST_PLATFORM" could not be found.
    + Query engine library for current platform "TEST_PLATFORM" could not be found.
      You incorrectly pinned it to TEST_PLATFORM

      This probably happens, because you built Prisma Client on a different platform.
      (Prisma Client looked in "/client/src/__tests__/integration/errors/missing-engine-native-binaryTarget/node_modules/@prisma/client/runtime/libquery_engine-TEST_PLATFORM.LIBRARY_TYPE.node")
```

File: packages/client/src/runtime/RequestHandler.ts [hunk 9]
```diff
@@ -154,37 +158,42 @@ export class RequestHandler {
...
-        const { stack } = printStack({
-          callsite,
-          originalMethod: clientMethod,
-          onUs: e.isPanic,
-          showColors,
-        })
-        message = `${stack}\n  ${e.message}`
-      }
+    } catch (error) {
+      this.handleRequestError({ error, clientMethod, callsite })
```
Comment (millsp) [inline packages/client/src/runtime/RequestHandler.ts:162]: Nice, thanks for the cleanup.

File: packages/client/src/runtime/dmmf-types.ts [hunk 10]
```diff
@@ -1,3 +1,5 @@
 import { DMMF } from '@prisma/generator-helper'
 
 export { DMMF }
+
+export type BundledDMMF = Pick<DMMF.Document, 'datamodel' | 'mappings'>
```
Comment (millsp) [inline packages/client/src/runtime/dmmf-types.ts:5]: I have called that "BaseDMMF", I think it would be good to align/rename all instances to be the same. I'm happy with either "bundled" or "base".

Comment (aqrln) [reply packages/client/src/runtime/dmmf-types.ts:5]: `BaseDMMF` sounds good

File: packages/client/src/runtime/RequestHandler.ts (hunk 9 above)
Comment (aqrln) [reply packages/client/src/runtime/RequestHandler.ts:162]: See [this commit](https://github.com/prisma/prisma/pull/13811/commits/7ad50cb785f124a156628363b6c13c2ce5ae9bd6) and [this comment](https://github.com/prisma/prisma/pull/13811#issuecomment-1159798372) for details.

**Review by millsp — APPROVED** [review]

File: packages/client/src/runtime/RequestHandler.ts (hunk 9 above)
Comment (millsp) [reply packages/client/src/runtime/RequestHandler.ts:162]: Ah perfect. Thanks!

**Comment by Jolg42** [comment]
Posted in #prisma-ecosystem-tests

:warning: From what I can see, this PR landed in `3.16.0-dev.82` and since then many tests are failing
see https://github.com/prisma/ecosystem-tests/actions/runs/2541422355

Failures seem to relate to this only
```
Error: node_modules/@prisma/client/runtime/index.d.ts(4,19): error TS2307: Cannot find module 'ts-toolbelt' or its corresponding type declarations.
```

Also in optional-test it broke the community-generators (typegraphql-prisma) test
https://github.com/prisma/ecosystem-tests/runs/7001616435?check_suite_focus=true#step:6:106
```
TypeError: Cannot read property 'enumTypes' of undefined
    at new DmmfDocument (/home/runner/work/ecosystem-tests/ecosystem-tests/community-generators/typegraphql-prisma/node_modules/typegraphql-prisma/lib/generator/dmmf/dmmf-document.js:10:30)
    at generateCode (/home/runner/work/ecosystem-tests/ecosystem-tests/community-generators/typegraphql-prisma/node_modules/typegraphql-prisma/lib/generator/generate-code.js:50:26)
    at Object.generate [as onGenerate] (/home/runner/work/ecosystem-tests/ecosystem-tests/community-generators/typegraphql-prisma/node_modules/typegraphql-prisma/lib/cli/prisma-generator.js:47:39)
    at async LineStream.<anonymous> (/home/runner/work/ecosystem-tests/ecosystem-tests/community-generators/typegraphql-prisma/node_modules/@prisma/generator-helper/dist/generatorHandler.js:33:24)
```
